### PR TITLE
H3 remake anvy

### PR DESCRIPTION
**Pull request by anastassiavybornova**
adding a remake of the workflow to
* divide study area into h3 grid cells (user defined resolution)
* compute metrics grid cell wise
* viz of overall metric ratios 
* viz single-cell plots (adding some examples, can be removed later)
this addresses (but does not completely resolve yet) issue #37 (feature matching/displacement still missing)

happy to hear your thoughts on this!
also, i'm pretty sure the cellwise metrics computation can be made much more efficient than it is now (cf the notebook cell with
```
# geom counts in cell
grid["geom_count_base"] = grid.apply(lambda x: geom_count(base, x.geometry), axis = 1)
```
etc., happy to hear suggestions

### CODE REVIEW

File: code/utils.py [hunk 1]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+
+def read_parenx(fua, option, proj_crs):
+    gdf = geopandas.read_parquet(f"../data/{fua}/parenx/{option}.parquet")
+    gdf = gdf.explode(index_parts=False)
+    gdf = gdf.reset_index(drop=True)
+    gdf = gdf.to_crs(proj_crs)
+    return gdf
+
+
+# singe-cell plot
```
Comment (jGaboardi) [inline code/utils.py:202]:
```suggestion
# single-cell plot
```

**Review by jGaboardi — APPROVED** [review]
Really cool stuff. Couple of minor comments. Nothing critical that much change for merge though. I can do a "jame" round of code tidying later on after merge (if desired).

File: code/utils.py [hunk 2]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    ]  # reverse coords
+    hex_pd = pd.DataFrame(list(h3.polyfill(temp, res=res)), columns=["hex_id"])
+    hex_pd["geometry"] = [
+        Polygon(h3.h3_to_geo_boundary(x, geo_json=True)) for x in hex_pd["hex_id"]
+    ]
+    grid = geopandas.GeoDataFrame(hex_pd)
+    grid.set_crs(orig_crs, inplace=True)
+
+    # keep only the grid cells that contain some piece of the orig data
+    mytree = strtree.STRtree(geoms=orig.geometry)
```
Comment (martinfleis) [inline code/utils.py:177]: Why not using `mytree = orig.sindex` instead of building it via shapely yourself?

File: code/utils.py [hunk 3]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    q = mytree.query(grid.geometry, predicate="intersects")
+    grid = grid.loc[sorted(set(q[0]))]
```
Comment (martinfleis) [inline code/utils.py:179]: geopandas `.sindex` has a `sort` keyword in the `query` method, so you don't need to do that yourself.

File: code/utils.py [hunk 4]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    grid = grid.reset_index(drop=True)
+    grid = grid.to_crs(proj_crs)
+    return grid
+
+
+def read_manual(fua, proj_crs):
+    gdf = geopandas.read_parquet(f"../data/{fua}/manual/{fua}.parquet")
+    gdf = gdf[["geometry"]]
+    gdf = gdf.explode(index_parts=False)
+    gdf = gdf.reset_index(drop=True)
```
Comment (martinfleis) [inline code/utils.py:189]:
```suggestion
    gdf = gdf.explode(ignore_index=True)
```

Same output.

File: code/utils.py [hunk 5]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    gdf = gdf.to_crs(proj_crs)
+    return gdf
+
+
+def read_parenx(fua, option, proj_crs):
+    gdf = geopandas.read_parquet(f"../data/{fua}/parenx/{option}.parquet")
+    gdf = gdf.explode(index_parts=False)
+    gdf = gdf.reset_index(drop=True)
```
Comment (martinfleis) [inline code/utils.py:197]:
```suggestion
    gdf = gdf.explode(ingore_index=True)
```

File: code/utils.py [hunk 6]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    fua: int, ID of city (any of use cases: 809, 869, 1133, 1656, 4617, 4881, 8989)
+    res: int, hex grid resolution
+    (suggested: 8 - with cell side length ~530m or 9 - cell side length ~200m)
+    """
+
+    # CRS for h3 must be 4326
+    orig_crs = "EPSG:4326"
+
+    # get city geom and name
+    meta = geopandas.read_parquet("../data/sample.parquet")
```
[the remark this answers is not in the document]
Comment (anastassiavybornova) [reply code/utils.py]: updated utils & notebook

File: code/utils.py [hunk 7]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    geom = meta.loc[meta.eFUA_ID == fua, "geometry"]
+
+    # read in OSM data
+    orig = geopandas.read_parquet(f"../data/{fua}/roads_osm.parquet")
```
[the remark this answers is not in the document]
Comment (anastassiavybornova) [reply code/utils.py]: updated utils & notebook

File: code/utils.py [hunk 8]
```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
...
+    temp = mapping(geom)["features"][0]["geometry"]
+    temp["coordinates"] = [
+        [[j[1], j[0]] for j in i] for i in temp["coordinates"]
+    ]  # reverse coords
+    hex_pd = pd.DataFrame(list(h3.polyfill(temp, res=res)), columns=["hex_id"])
+    hex_pd["geometry"] = [
+        Polygon(h3.h3_to_geo_boundary(x, geo_json=True)) for x in hex_pd["hex_id"]
+    ]
+    grid = geopandas.GeoDataFrame(hex_pd)
+    grid.set_crs(orig_crs, inplace=True)
```
[the remark this answers is not in the document]
Comment (anastassiavybornova) [reply code/utils.py]: great to know. i always found this inplace stuff sneaky. replaced it

File: code/utils.py (hunk 2 above)
Comment (anastassiavybornova) [reply code/utils.py]: changed it, thanks for the hint

File: code/utils.py (hunk 3 above)
Comment (anastassiavybornova) [reply code/utils.py]: ... and likewise

**Comment by anastassiavybornova** [comment]
hi! @martinfleis  and @jGaboardi i've implemented all of your changes suggested above, plus the clipping only once-per-apply func as discussed yesterday. from my side it's ready to merge, and @jGaboardi you're obviously (as hinted above) always welcome to do your james stuff on it.

**Comment by jGaboardi** [comment]
My approval still stands, but we'll wait to see if @martinfleis has any more comments.

File: code/utils.py [hunk 9]
```diff
@@ -1,12 +1,15 @@
 import json
 import pathlib
+from collections import Counter
 
 import contextily as cx
 import cv2
 import geopandas
+import matplotlib.gridspec as gridspec
```
Comment (martinfleis) [inline code/utils.py:8]:
```suggestion
from matplotlib import gridspec
```

@jGaboardi any idea why Ruff did not complain about this? Normally it does? Maybe some rules to allow?

**Review by martinfleis — APPROVED** [review]
I _think_ that you are not using the pre-commit hook @jGaboardi set but apart from that, all good.

File: code/utils.py (hunk 9 above)
Comment (jGaboardi) [reply code/utils.py:8]: I do not know... but I have noticed recently some instance in my work stuff where `ruff` is silent on things that it should be complaining about. Perhaps it's a configuration difference that has changed in the latest `ruff` releases.

**Comment by jGaboardi** [comment]
> I _think_ that you are not using the pre-commit hook @jGaboardi set but apart from that, all good.

@anastassiavybornova Can you run `ruff` manually on the `code/` directory to see if it complains?

```
ruff check code
```

**Comment by anastassiavybornova** [comment]
hmmm that's weird, i am using the pre commit hooks and i actually did run the manual `ruff` before committing,

```
(simplification) anvy@mac622265  simplification % ruff check code/utils.py      
All checks passed!
```

**Comment by anastassiavybornova** [comment]
`ruff check code/` does throw a bunch of errors for four of the notebooks (evaluate_h3cells, momepy, parenx, and _add_artifacts), but I think that's not what you were referring to @martinfleis ?

**Comment by martinfleis** [comment]
Somehow pre-commit is passing but ruff check complains. Not sure why. I fixed some bits, so let's merge this. We can look into the pre-commit issue later.

**Pull request merged**